### PR TITLE
pageserver: use BufferedWriter (and direct IO if enabled) for delta + image layer writers

### LINKED ISSUE #9868: pageserver: direct IO on the write path
### Tasks
- https://github.com/neondatabase/neon/pull/11543
- https://github.com/neondatabase/neon/pull/11585
- https://github.com/neondatabase/neon/pull/11544
- https://github.com/neondatabase/neon/pull/10063
- https://github.com/neondatabase/neon/pull/11667
- https://github.com/neondatabase/neon/pull/11549
- https://github.com/neondatabase/neon/pull/11558
- https://github.com/neondatabase/infra/pull/3536
- https://github.com/neondatabase/infra/pull/3566
- https://github.com/neondatabase/infra/pull/3616
- https://github.com/neondatabase/neon/issues/11762
- https://github.com/neondatabase/neon/issues/11598
- https://github.com/neondatabase/neon/issues/11676
- https://github.com/neondatabase/infra/pull/3664
- https://github.com/neondatabase/infra/pull/3695
- https://github.com/neondatabase/infra/pull/3700
- https://github.com/neondatabase/infra/pull/3778

### Punted
- https://github.com/neondatabase/neon/issues/11692
- https://github.com/neondatabase/neon/issues/11817

### PR DESCRIPTION
**Pull request by yliang412**
closes #9868 

## Problem

Previously delta and image layer writes are using buffered io and written to disk using `BlobWriter`, which uses an internal buffer for delta layers, and bypass the internal buffer for image layers. To enable O_DIRECT on the write path, we need to refactor the code to do aligned writes.

## Summary of changes

- `BlobWriter` now uses double-buffered writer implemented in https://github.com/neondatabase/neon/pull/9693 to do writes. Previously, it is directly using `VirtualFile` + internal buffer.
- Timeline gate is propagated everywhere `BufferedWriter` is used to properly hold the background flush task open.
- BufferedWriter `Drop` handling:
  - use monotonically increasing number instead of random string for delta/image layer tempfile suffix.
  - spawn a tokio task to join the background flush task and drop inner `VirtualFile`.

### Alignment Requirement
 
buffer alignment: guaranteed by the aligned allocation of `IoBufferMut`.

There are two cases to guarantee file offset alignment: 
  - [x] buffer is full: the buffer capacity is a multiple of DIO alignment (512), and start offset is always a multiple of DIO alignment.
  - [x] buffer is not full: we always pad zeros to the next multiple of DIO alignment (512).
    - `InMemoryLayer` did not have this problem when we merged https://github.com/neondatabase/neon/pull/9693 because the tail of the reads is always kept in memory.
    - Not fragmentation issue because this only happens to the last write.
    - offsets are tracked in index layers, and length headers are stored in the blobs. We are not relying on blobs being next to each other to calculate length.

In addition, the length of summary and index blocks always `PAGE_SZ` and aligned to DIO requirement (512). The IOs for these blocks are directly performed by through `VirtualFile`.

### CODE REVIEW

**Comment by yliang412** [comment]
CI with O_DIRECT timed out on a few of the tests: https://neon-github-public-dev.s3.amazonaws.com/reports/pr-10063/12264248904/index.html

File: pageserver/src/tenant/storage_layer/delta_layer.rs [hunk 1]
```diff
@@ -725,7 +736,7 @@ impl Drop for DeltaLayerWriter {
             // We want to remove the virtual file here, so it's fine to not
             // having completely flushed unwritten data.
             let vfile = inner.blob_writer.into_inner_no_flush();
-            vfile.remove();
+            std::fs::remove_file(vfile.path()).expect("failed to remove the virtual file");
```
Comment (problame) [inline pageserver/src/tenant/storage_layer/delta_layer.rs:739]: the .remove() function can be removed now, dead `pub` code.

File: pageserver/src/tenant/storage_layer/image_layer.rs [hunk 2]
```diff
@@ -873,15 +884,16 @@ impl ImageLayerWriterInner {
         crate::metrics::COMPRESSION_IMAGE_INPUT_BYTES_CHOSEN.inc_by(self.uncompressed_bytes_chosen);
         crate::metrics::COMPRESSION_IMAGE_OUTPUT_BYTES.inc_by(compressed_size);
 
-        let mut file = self.blob_writer.into_inner();
+        let file = self.blob_writer.into_inner(ctx).await?;
 
         // Write out the index
-        file.seek(SeekFrom::Start(index_start_blk as u64 * PAGE_SZ as u64))
-            .await?;
+        // TODO(yuchen): should we just replace BlockBuf::blocks with one big buffer?
```
Comment (problame) [inline pageserver/src/tenant/storage_layer/image_layer.rs:890]: Don't see a reason why not. Please create a follow-up good_first_issue like `investigate & do if possible: replace BlockBuf::blocks with one big buffer`

File: pageserver/src/virtual_file/owned_buffers_io/aligned_buffer/buffer_mut.rs [hunk 3]
```diff
@@ -280,6 +280,17 @@ unsafe impl<A: Alignment> tokio_epoll_uring::IoBufMut for AlignedBufferMut<A> {
     }
 }
 
+impl<A: Alignment> std::io::Write for AlignedBufferMut<A> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.extend_from_slice(buf);
```
Comment (problame) [inline pageserver/src/virtual_file/owned_buffers_io/aligned_buffer/buffer_mut.rs:285]: Was going to flag that this may panic, but turns out the implementation doesn't.
Kinda confusing IMO, the convention from the `bytes` crate is that `extend_from_slice` does panic.
Didn't I flag this in some earlier review?

Maybe add a param `enum ResizeBehavior { Panic, Doit }` to make it explicit?

File: pageserver/src/virtual_file/owned_buffers_io/write/flush.rs [hunk 4]
```diff
@@ -193,6 +196,14 @@ where
         handle.join_handle.await.unwrap()
     }
 
+    pub fn shutdown_no_flush(mut self) {
+        let handle = self
+            .inner
+            .take()
+            .expect("must not use after we returned an error");
+        handle.join_handle.abort();
+    }
```
Comment (problame) [inline pageserver/src/virtual_file/owned_buffers_io/write/flush.rs:205]: So, this would be one of the first instances where we actually exercise the cancel-safety of tokio-epoll-uring.

Comment (problame) [reply pageserver/src/virtual_file/owned_buffers_io/write/flush.rs:205]: Actually this isn't safe. 

Our caller (`impl Drop for DeltaLayerWriter`) will delete the file from the file system.

But the flush task uses VirtualFile.

With VirtualFile, it can theoretically happen that the file descriptor cache has a miss and we need to re-open the file.

That reopen will either fail (good) or it could open a new inode at that path (bad).

Comment (problame) [reply pageserver/src/virtual_file/owned_buffers_io/write/flush.rs:205]: if instead of random suffix we would use strongly monotonically increasing numbers for the delta|image layer writer temp files, then I guess there would be no risk of the (bad) case above.

https://github.com/neondatabase/neon/blob/e05b5d6ae4a7dfd58ae0cfe62403a7ae340b3356/pageserver/src/tenant/storage_layer/delta_layer.rs#L302-L316

File: pageserver/src/virtual_file.rs [hunk 5]
```diff
@@ -975,7 +975,10 @@ impl VirtualFileInner {
     ) -> (FullSlice<B>, Result<usize, Error>) {
         let file_guard = match self.lock_file().await {
             Ok(file_guard) => file_guard,
-            Err(e) => return (buf, Err(e)),
+            Err(e) => {
+                println!("ERRORED :(");
```
Comment (problame) [inline pageserver/src/virtual_file.rs:979]: stray debugging code?

Comment (problame) [reply pageserver/src/virtual_file.rs:979]: please remove

also for future: `dbg!(e)`

File: pageserver/src/virtual_file.rs [hunk 6]
```diff
@@ -1329,7 +1332,8 @@ impl OwnedAsyncWriter for VirtualFile {
         ctx: &RequestContext,
     ) -> std::io::Result<FullSlice<Buf>> {
         let (buf, res) = VirtualFile::write_all_at(self, buf, offset, ctx).await;
-        res.map(|_| buf)
+        let x = res.map(|_| buf).unwrap();
+        Ok(x)
```
Comment (problame) [inline pageserver/src/virtual_file.rs:1336]: Why is unwrap ok?

**Review by problame — COMMENTED** [review]
Passing the gate through the RequestContext and making it clonable would have probably avoided a bunch of pain and function signature bloat?  cc @jcsp

---

The meat of the PR is in
- `blob_io.rs`
- `disk_btree.rs`
- `flush.rs`
- `write.rs`

Have reviewed all files except `blob_io.rs` and `disk_btree.rs`, need to hurry to a meeting now.

File: pageserver/src/virtual_file/owned_buffers_io/aligned_buffer/buffer_mut.rs (hunk 3 above)
Comment (yliang412) [reply pageserver/src/virtual_file/owned_buffers_io/aligned_buffer/buffer_mut.rs:285]: I think both `bytes::BytesMut::extend_from_slice` and `std::vec::Vec::extend_from_slice` resizes?

https://docs.rs/bytes/latest/bytes/struct.BytesMut.html#method.extend_from_slice

> If this BytesMut object does not have enough capacity, it is resized first.

https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend_from_slice

> Clones and appends all elements in a slice to the Vec.

All impls including our `IoBufferMut` could panic if the new capacity exceeds `isize::MAX` bytes.

Comment (yliang412) [reply pageserver/src/virtual_file/owned_buffers_io/aligned_buffer/buffer_mut.rs:285]: Are you referring to the fact that `bytes::BytesMut` does not itself actually implement `std::io::Write` and we are relying on `&mut [u8]`? Then in that case, yea we would panic if we did not reserve enough bytes.

Also it seems the proper way to use `std::io::Write` with `BytesMut` is through the `Writer<B> where B: BufMut` API.

https://docs.rs/bytes/latest/bytes/buf/struct.Writer.html#impl-Write-for-Writer%3CB%3E

Comment (problame) [reply pageserver/src/virtual_file/owned_buffers_io/aligned_buffer/buffer_mut.rs:285]: No, somehow I was under the impression that `extend_from_slice` doesn't resize in most APIs but panics instead.

File: pageserver/src/virtual_file.rs (hunk 6 above)
Comment (yliang412) [reply pageserver/src/virtual_file.rs]: 4b358f6a3c60b809aa2ff0a2672aa20a5197368a: properly error handling by checking `res?`

File: pageserver/src/virtual_file.rs (hunk 5 above)
Comment (yliang412) [reply pageserver/src/virtual_file.rs]: 3d4052f253f9b1f6233d670753248fa8ef4bac41

File: pageserver/src/tenant/storage_layer/image_layer.rs (hunk 2 above)
Comment (yliang412) [reply pageserver/src/tenant/storage_layer/image_layer.rs]: daf214e2e9e8c04df11c75d2fd1c1f603f44f9ec: created issue at https://github.com/neondatabase/neon/issues/10092

File: pageserver/src/virtual_file/owned_buffers_io/write/flush.rs (hunk 4 above)
Comment (yliang412) [reply pageserver/src/virtual_file/owned_buffers_io/write/flush.rs:205]: > With VirtualFile, it can theoretically happen that the file descriptor cache has a miss and we need to re-open the file.

In the case you mentioned, since we set `create` options to `false` for the `reopen_options`, I think we are good? The flush task will fail to reopen the file the file has been removed. 

https://github.com/neondatabase/neon/blob/7fa986bc923b6af3c77e888be96123f7ba068313/pageserver/src/virtual_file.rs#L612-L620.

Comment (problame) [reply pageserver/src/virtual_file/owned_buffers_io/write/flush.rs:205]: > The flush task will fail to reopen the file the file has been removed.

Yeah, the flush task will exit at some later point after the drop (not under our control).

But what if in the meantime the path is occupied by a new file.

There is a schedule where the flush task _could_ would then write into that new file, and that would be corruption:

```
flush task: returns from recv.await with some flush work
T1: drops FlushHandle
T1: std::fs::remove()'s the file from the filesystem
T1: creates a new file with same name in the filesystem (same layer file name and same `rand_string`)
flush task: tries to do the flush work through `VirtualFile::write_...`
flush_task: the VirtualFile::write_ needs to do reopen
flush_task: the VirtualFile::write_ writes into the new file
```

I think if we replace the `rand_string` with a monotonic counter, like we do with EphemeralFile names, we eliminate this case.

Basically it was always the implicit requirement of VirtualFile that

* either you make sure all the VirtualFile instances referencing a path are dropped before you delete it (now-unused `VirtualFile::remove`. API guarantees it by consuming `self`)
* or you disambiguate through the file name, then it's safe to delete a path for which there are still VirtualFile objects in existence, these will simply encounter errors and those need to be dealt with.

The use of `Arc<VirtualFile>` kind of puts ourselves into a corner where we have to do both.

Comment (problame) [reply pageserver/src/virtual_file/owned_buffers_io/write/flush.rs:205]: Actually we can just spawn the shutdown into a background tokio task.

You only need to keep the Timeline gate open once more.

```rust
fn drop(&mut self) {
   let Some(inner) = self.inner.take() else {
     return;
   };
   tokio::spawn(async move {
     let vfile = inner.blob_writer.shutdown().await;
     // it's ok to use sync io for namespace operations
     // https://github.com/neondatabase/neon/issues/7370
     if let Err(e) = std::fs::remove_file(vfile.path())  {
       e.maybe_fatal_err();
       // nothing we can do about it at this point
       error!(err=%e, path=%vfile.path(), "failed to remove layer writer file");
     }
     // the task keeps the timeline gate open until it's done operating on the timeline dir
     let Inner { gate_guard, .. } = { inner };
     drop(gate_guard);
   })
}
```

**Review by problame — CHANGES_REQUESTED** [review]
`blob_io.rs` is a breath of fresh air - nice cleanup!

---

I think my only hard objection in addition to the ongoing conversation around how we deal with dropping of the layer writers of course, is how you do the padding to next multiple during `into_inner` / `shutdown` . That seem incorrect (B::ALIGN) and in the wrong place to me.

---

In terms of perf testing:

1. bench_ingest is prob good enough as a first impression
2. image layer writer isn't covered by bench_ingest ; I think it's fine to copy-paste bench_ingest and trim it down to a tiny image  layer writer microbenchmark; 8k, zeroes and random values; 
3. image layer writer was writing the images directly to kernel page cache before this PR, that's no longe rthe case; I would like to see a comparison of `before-with-buffered,after-with-buffered,after-with-direct.

File: pageserver/src/virtual_file/owned_buffers_io/write.rs [hunk 7]
```diff
@@ -107,31 +109,59 @@ where
...
-        self.flush(ctx).await?;
+    pub async fn shutdown(mut self, ctx: &RequestContext) -> std::io::Result<(u64, W)> {
+        let buf = self.tail_mut();
+        let len = buf.pending();
+        let cap = buf.cap();
+        if len < cap {
+            // pad zeros to the next io alignment requirement.
+            let count = len.next_multiple_of(B::ALIGN).min(cap) - len;
+            buf.extend_with(0, count);
+        }
```
[the remark this answers is not in the document]
Comment (yliang412) [reply pageserver/src/virtual_file/owned_buffers_io/write.rs:120]: Direct IO alignment requirement:

- buffer at aligned memory address.
- file offset to write at is aligned.
- buffer size written out is multiple of alignment.

Without the padding, `VirtualFile::write_all_at` fails with `invalid argument` because the buffer size we are writing out is potentially not a multiple of 512.

File: pageserver/src/virtual_file/owned_buffers_io/write/flush.rs (hunk 4 above)
Comment (yliang412) [reply pageserver/src/virtual_file/owned_buffers_io/write/flush.rs]: b1edbf667c61d842db0b7c4076d0ae98433b1128 adds the monotonically increasing file suffix.
38f1150ae29bda4952c52e0bdae9346e21e78a69 use the background task for dropping BlobWriter.

File: pageserver/src/virtual_file/owned_buffers_io/write.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (yliang412) [reply pageserver/src/virtual_file/owned_buffers_io/write.rs]: 38f1150ae29bda4952c52e0bdae9346e21e78a69

**Comment by yliang412** [comment]
Didn't get to benchmark

**Comment by problame** [comment]
Made the unit tests pass; failures were largely due to problems I spotted while reviewing ☺️

image

**Comment by problame** [comment]
I decided to leave it alone after rebasing and adding TODOs for things I spotted along the way. The strategy is to extract preliminary refactorings + a final PR.
My output last&this week are these preliminary refactorings here:
- https://github.com/neondatabase/neon/pull/11585
- https://github.com/neondatabase/neon/pull/11543
- https://github.com/neondatabase/neon/pull/11549

The big rock that implements the behavior from this PR + more is the:
- https://github.com/neondatabase/neon/pull/11558
